add_source_file

```diff
@@ -200,18 +200,70 @@ def add_to_code_blocks_target(filename, target):
             break
         index += 1
     cbp_lines.insert(index, elem)
-    
+
     open(cbp_file, 'w').writelines(cbp_lines)
 
 def add_to_code_blocks(filename, targets):
     translated_targets = [code_blocks_target_translations[t] for t in targets]
     print(" code::blocks targets:", translated_targets)
     for t in translated_targets:
         add_to_code_blocks_target(filename, t)
-        # if there's also an .hpp file, add that too
-        hpp = filename.with_suffix(".hpp")
-        if rootdir.joinpath("src", hpp).exists():
-            add_to_code_blocks_target(hpp, t)
+
+def sanity_check_existing_cpp_hpp(filenames):
+    """
+    If we're adding a .cpp file, check whether a .hpp should be added too, etc.
+    Only the files named on the command line are added, this exits if the check fails.
+    """
+    any_check_failed = False
+    for filename in filenames:
+        if filenames.count(filename) > 1:
+            print(f"ERROR: File '{filename}' given multiple times")
+            any_check_failed = True
+
+        if not rootdir.joinpath("src", filename).exists():
+            print(f"WARN: File '{filename}' does not exist")
+            any_check_failed = True
+
+        spouse = None
+        if filename.suffix == ".cpp":
+            spouse = filename.with_suffix(".hpp")
+        elif filename.suffix == ".hpp":
+            spouse = filename.with_suffix(".cpp")
+
+        if rootdir.joinpath("src", spouse).exists() and not filenames.count(spouse):
+            print(f"WARN: Requested to add '{filename}', should '{spouse}' be added too?")
```

Comment:
If it's going to force a failure, then it's not really a question, is it?

Reply:
It'll also suggest using the `--no-checks` option, I'm wanting the user to think about which of the two possibilities is the correct one; it might be a case where the .hpp file has already been added, but the .cpp file is a new addition.